core/src/main/java/com/linecorp/armeria/internal/common/DefaultRequestTarget.java

```diff
@@ -90,158 +80,229 @@ public final class PathAndQuery {
-            Flags.parsedPathCacheSpec() != null ? buildCache(Flags.parsedPathCacheSpec()) : null;
-
-    private static Cache<String, PathAndQuery> buildCache(String spec) {
-        return Caffeine.from(spec).build();
-    }
-
-    public static void registerMetrics(MeterRegistry registry, MeterIdPrefix idPrefix) {
-        if (CACHE != null) {
-            CaffeineMetricSupport.setup(registry, idPrefix, CACHE);
-        }
-    }
+    private static final Bytes EMPTY_BYTES = new Bytes(0);
+    private static final Bytes SLASH_BYTES = new Bytes(new byte[] { '/' });
 
     /**
-     * Clears the currently cached parsed paths. Only for use in tests.
+     * The main implementation of {@link RequestTarget#forServer(String)}.
      */
-    @VisibleForTesting
-    public static void clearCachedPaths() {
-        requireNonNull(CACHE, "CACHE");
-        CACHE.asMap().clear();
-    }
+    @Nullable
+    public static RequestTarget forServer(String reqTarget, boolean allowDoubleDotsInQueryString) {
+        final RequestTarget cached = RequestTargetCache.getForServer(reqTarget);
+        if (cached != null) {
+            return cached;
+        }
 
-    /**
-     * Returns paths that have had their parse result cached. Only for use in tests.
-     */
-    @VisibleForTesting
-    public static Set<String> cachedPaths() {
-        requireNonNull(CACHE, "CACHE");
-        return CACHE.asMap().keySet();
+        return slowForServer(reqTarget, allowDoubleDotsInQueryString);
     }
 
     /**
-     * Validates the {@link String} that contains an absolute path and a query, and splits them into
-     * the path part and the query part. If the path is usable (e.g., can be served a successful response from
-     * the server and doesn't have variable path parameters), {@link PathAndQuery#storeInCache(String)} should
-     * be called to cache the parsing result for faster future invocations.
-     *
-     * @return a {@link PathAndQuery} with the absolute path and query, or {@code null} if the specified
-     *         {@link String} is not an absolute path or invalid.
+     * The main implementation of {@link RequestTarget#forClient(String, String)}.
      */
     @Nullable
-    public static PathAndQuery parse(@Nullable String rawPath) {
-        return parse(rawPath, Flags.allowDoubleDotsInQueryString());
-    }
-
-    @VisibleForTesting
-    @Nullable
-    static PathAndQuery parse(@Nullable String rawPath, boolean allowDoubleDotsInQueryString) {
-        if (CACHE != null && rawPath != null) {
-            final PathAndQuery parsed = CACHE.getIfPresent(rawPath);
-            if (parsed != null) {
-                return parsed;
+    public static RequestTarget forClient(String reqTarget, @Nullable String prefix) {
+        requireNonNull(reqTarget, "reqTarget");
+
+        final int authorityPos = findAuthority(reqTarget);
+        if (authorityPos >= 0) {
+            final RequestTarget cached = RequestTargetCache.getForClient(reqTarget);
```

Comment:
Question: Shouldn't we need to take `prefix` into account to get the cached value?

Reply:
An absolute URI doesn't use `prefix` at all and thus it's ignored as intended. Let me add some comment.